### PR TITLE
Add new API Key authenticator impl

### PR DESCRIPTION
**Pull request by mevan-karu**
### Purpose
$subject

Replaces the temp authenticator added with https://github.com/wso2/product-microgateway/pull/3603

### Issues

Fixes #

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [ ] Assigned 'Type' label
- [ ] Assigned the project
- [ ] Validated respective github issues
- [ ] Assigned milestone to the github issue(s)

### CODE REVIEW

**Review by senthalan — APPROVED** [review]

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyAuthenticator.java [hunk 1]
```diff
@@ -55,17 +50,34 @@ public APIKeyAuthenticator() {
...
+        AuthenticationContext authCtx = super.authenticate(requestContext);
+        // Drop the API key data from the API key header.
+        dropAPIKeyDataFromAPIKeyHeader(requestContext);
+        return authCtx;
+    }
+
+    private void dropAPIKeyDataFromAPIKeyHeader(RequestContext requestContext) throws APISecurityException {
+
+        String apiKeyHeaderValue = getAPIKeyFromRequest(requestContext).trim();
+        String checksum = apiKeyHeaderValue.substring(apiKeyHeaderValue.length() - 6);
```
Comment (VirajSalaka) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyAuthenticator.java:69]: Should we check the length and do processing as indexOutOfBound exception could be thrown.

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java [hunk 2]
```diff
@@ -557,7 +557,7 @@ public int getPriority() {
      * @param jwtToken          JWT Token
      * @throws APISecurityException in case of scope validation failure
      */
-    private void validateScopes(String apiContext, String apiVersion, ResourceConfig matchingResource,
+    protected void validateScopes(String apiContext, String apiVersion, ResourceConfig matchingResource,
```
Comment (VirajSalaka) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java:560]: This access modifier not required?

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyAuthenticator.java (hunk 1 above)
Comment (VirajSalaka) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyAuthenticator.java:69]: I guess we can process this in can authenticate method straightaway and then no need to check that repeatedly everywhere.

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyAuthenticator.java [hunk 3]
```diff
@@ -74,26 +86,50 @@ private String getAPIKeyFromRequest(RequestContext requestContext) {
...
+    @Override
+    protected String retrieveTokenFromRequestCtx(RequestContext requestContext) throws APISecurityException {
+
+        String apiKey = getAPIKeyFromRequest(requestContext);
+        if (!APIKeyUtils.isValidAPIKey(apiKey)) {
+            throw new APISecurityException(APIConstants.StatusCodes.UNAUTHENTICATED.getCode(),
+                    APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
+                    APISecurityConstants.API_AUTH_INVALID_CREDENTIALS_MESSAGE);
+        }
+        String keyHash = APIKeyUtils.generateAPIKeyHash(apiKey);
```
Comment (VirajSalaka) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyAuthenticator.java:113]: This method also could generate IllegalArgumentException if not a valid base64 payload.

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/api/Utils.java [hunk 4]
```diff
@@ -92,6 +92,13 @@ static void populateRemoveAndProtectedHeaders(RequestContext requestContext) {
             return;
         }
 
+        // Choreo-API-Key is considered as a protected header, hence header value should be treated
+        // same as other security headers.
+        if (ConfigHolder.getInstance().getConfig().getApiKeyConfig().getApiKeyInternalHeader() != null) {
```
[the remark this answers is not in the document]
Comment (mevan-karu) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/api/Utils.java:97]: A default config is set for this from the adapter [1].

[1] https://github.com/wso2/product-microgateway/blob/choreo/adapter/config/default_config.go#L231

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyAuthenticator.java [hunk 5]
```diff
@@ -55,17 +50,34 @@ public APIKeyAuthenticator() {
...
-        }
         String apiKeyValue = getAPIKeyFromRequest(requestContext);
         return apiKeyValue != null && apiKeyValue.startsWith(APIKeyConstants.API_KEY_PREFIX);
     }
 
     @Override
     public AuthenticationContext authenticate(RequestContext requestContext) throws APISecurityException {
 
-        return super.authenticate(requestContext);
+        AuthenticationContext authCtx = super.authenticate(requestContext);
```
[the remark this answers is not in the document]
Comment (mevan-karu) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyAuthenticator.java:60]: It's not required as any sort of failure in `super.authenticate()` will not forward the request to the downstream service.

**Review by VirajSalaka — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[failed] Dataplane(EastUS) cluster : dev-deployment-v2 : 20241029.30

**Comment by choreo-cicd** [comment]
[failed]  : dev-deployment-v2 : 20241029.30

**Comment by choreo-cicd** [comment]
[failed] Dataplane(EastUS) cluster : dev-deployment-v2 : 20241029.31

**Comment by choreo-cicd** [comment]
[failed]  : dev-deployment-v2 : 20241029.31

**Comment by choreo-cicd** [comment]
[failed] Dataplane(EastUS) cluster : dev-deployment-v2 : 20241029.32

**Comment by choreo-cicd** [comment]
[failed] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20241029.32

**Comment by choreo-cicd** [comment]
[failed] Dataplane(EastUS) cluster : dev-deployment-v2 : 20241029.33

**Comment by choreo-cicd** [comment]
[failed] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20241029.33

**Comment by choreo-cicd** [comment]
[failed]  : dev-deployment-v2 : 20241029.33

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20241029.34

**Comment by choreo-cicd** [comment]
[] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20241030.32

**Comment by choreo-cicd** [comment]
[]  : dev-deployment-v2 : 20241030.32

**Comment by choreo-cicd** [comment]
[] Dataplane(EastUS) cluster : dev-deployment-v2 : 20241030.32

**Comment by choreo-cicd** [comment]
[] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20241031.3

**Comment by choreo-cicd** [comment]
[] Dataplane(EastUS) cluster : stage-deployment-v2 : 20241031.3

**Comment by choreo-cicd** [comment]
[] Controlplane cluster : stage-deployment-v2 : 20241031.3

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : prod-deployment-v2 : 20241101.3

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20241101.3

**Comment by choreo-cicd** [comment]
[failed] Dataplane(CDP2-WU2) cluster : prod-deployment-v2 : 20241101.3

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : prod-deployment-v2 : 20241101.3

**Comment by choreo-cicd** [comment]
[] Dataplane(CDP2-WU2) cluster : prod-deployment-v2 : 20241101.4

**Comment by choreo-cicd** [comment]
[] Dataplane(EastUS) cluster : prod-deployment-v2 : 20241101.4

**Comment by choreo-cicd** [comment]
[] Controlplane cluster : prod-deployment-v2 : 20241101.4

**Comment by choreo-cicd** [comment]
[] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20241101.4

**Pull request merged**